File: android/app/src/main/java/com/alyssa/polaris/LocalDataSqlitePlugin.java
package com.alyssa.polaris;

import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteStatement;
import android.util.Base64;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.io.File;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Locale;
import java.util.Set;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import org.json.JSONArray;
import org.json.JSONObject;

@CapacitorPlugin(name = "LocalDataSqlite")
public class LocalDataSqlitePlugin extends Plugin {
    private final ExecutorService sqliteExecutor = Executors.newSingleThreadExecutor();
    private SQLiteDatabase database;

    private static final String CREATE_ENTRIES_TABLE_SQL =
        "CREATE TABLE IF NOT EXISTS local_data_entries (" +
        " key TEXT PRIMARY KEY NOT NULL," +
        " value_json TEXT NOT NULL," +
        " updated_at INTEGER NOT NULL" +
        ")";
    private static final String READ_ENTRY_SQL =
        "SELECT value_json FROM local_data_entries WHERE key = ? LIMIT 1";
    private static final String LIST_KEYS_WITH_PREFIX_SQL =
        "SELECT key FROM local_data_entries WHERE substr(key, 1, ?) = ?";
    private static final String UPSERT_ENTRY_SQL =
        "INSERT INTO local_data_entries (key, value_json, updated_at) " +
        "VALUES (?, ?, ?) " +
        "ON CONFLICT(key) DO UPDATE SET " +
        "value_json = excluded.value_json, " +
        "updated_at = excluded.updated_at";
    private static final String DELETE_ENTRY_SQL =
        "DELETE FROM local_data_entries WHERE key = ?";
    private static final String CREATE_CHAT_CONVERSATION_TABLE_SQL =
        "CREATE TABLE IF NOT EXISTS chat_conversation (" +
        " id TEXT PRIMARY KEY NOT NULL," +
        " title TEXT NOT NULL," +
        " kind TEXT NOT NULL," +
        " collaborator_id TEXT," +
        " group_room_id TEXT," +
        " active_project_id TEXT," +
        " pinned_at INTEGER," +
        " created_at INTEGER NOT NULL," +
        " updated_at INTEGER NOT NULL," +
        " metadata_json TEXT NOT NULL" +
        ")";
    private static final String CREATE_CHAT_MESSAGE_TABLE_SQL =
        "CREATE TABLE IF NOT EXISTS chat_message (" +
        " id TEXT PRIMARY KEY NOT NULL," +
        " conversation_id TEXT NOT NULL," +
        " seq INTEGER NOT NULL," +
        " role TEXT NOT NULL," +
        " content TEXT NOT NULL," +
        " reasoning TEXT NOT NULL," +
        " created_at INTEGER NOT NULL," +
        " updated_at INTEGER NOT NULL," +
        " payload_json TEXT NOT NULL," +
        " FOREIGN KEY(conversation_id) REFERENCES chat_conversation(id) ON DELETE CASCADE," +
        " UNIQUE(conversation_id, seq)" +
        ")";
    private static final String CREATE_CHAT_CONVERSATION_UPDATED_INDEX_SQL =
        "CREATE INDEX IF NOT EXISTS idx_chat_conversation_updated_at " +
        "ON chat_conversation(updated_at DESC, id ASC)";
    private static final String CREATE_CHAT_MESSAGE_CONVERSATION_SEQ_INDEX_SQL =
        "CREATE INDEX IF NOT EXISTS idx_chat_message_conversation_seq " +
        "ON chat_message(conversation_id, seq)";
    private static final String UPSERT_CHAT_CONVERSATION_SQL =
        "INSERT INTO chat_conversation (" +
        " id," +
        " title," +
        " kind," +
        " collaborator_id," +
        " group_room_id," +
        " active_project_id," +
        " pinned_at," +
        " created_at," +
        " updated_at," +
        " metadata_json" +
        ") VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?) " +
        "ON CONFLICT(id) DO UPDATE SET " +
        "title = excluded.title, " +
        "kind = excluded.kind, " +
        "collaborator_id = excluded.collaborator_id, " +
        "group_room_id = excluded.group_room_id, " +
        "active_project_id = excluded.active_project_id, " +
        "pinned_at = excluded.pinned_at, " +
        "created_at = excluded.created_at, " +
        "updated_at = excluded.updated_at, " +
        "metadata_json = excluded.metadata_json";
    private static final String DELETE_CHAT_CONVERSATION_MESSAGES_SQL =
        "DELETE FROM chat_message WHERE conversation_id = ?";
    private static final String UPSERT_CHAT_MESSAGE_SQL =
        "INSERT INTO chat_message (" +
        " id," +
        " conversation_id," +
        " seq," +
        " role," +
        " content," +
        " reasoning," +
        " created_at," +
        " updated_at," +
        " payload_json" +
        ") VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?) " +
        "ON CONFLICT(id) DO UPDATE SET " +
        "conversation_id = excluded.conversation_id, " +
        "seq = excluded.seq, " +
        "role = excluded.role, " +
        "content = excluded.content, " +
        "reasoning = excluded.reasoning, " +
        "created_at = excluded.created_at, " +
        "updated_at = excluded.updated_at, " +
        "payload_json = excluded.payload_json";
    private static final String READ_CHAT_CONVERSATION_SUMMARIES_SQL =
        "SELECT " +
        "c.id, " +
        "c.title, " +
        "c.kind, " +
        "c.collaborator_id, " +
        "c.group_room_id, " +
        "c.active_project_id, " +
        "c.pinned_at, " +
        "c.created_at, " +
        "c.updated_at, " +
        "COUNT(m.id) AS message_count, " +
        "COALESCE(MAX(m.updated_at), 0) AS latest_message_timestamp " +
        "FROM chat_conversation c " +
        "LEFT JOIN chat_message m ON m.conversation_id = c.id " +
        "GROUP BY " +
        "c.id, " +
        "c.title, " +
        "c.kind, " +
        "c.collaborator_id, " +
        "c.group_room_id, " +
        "c.active_project_id, " +
        "c.pinned_at, " +
        "c.created_at, " +
        "c.updated_at " +
        "ORDER BY c.updated_at DESC, c.id ASC";
    private static final String READ_CHAT_CONVERSATION_EXISTS_SQL =
        "SELECT id FROM chat_conversation WHERE id = ? LIMIT 1";
    private static final String READ_CHAT_CONVERSATION_METADATA_SQL =
        "SELECT metadata_json FROM chat_conversation WHERE id = ? LIMIT 1";
    private static final String READ_CHAT_MESSAGE_COUNT_SQL =
        "SELECT COUNT(*) AS message_count FROM chat_message WHERE conversation_id = ?";
    private static final String READ_RECENT_CHAT_MESSAGES_SQL =
        "SELECT seq, payload_json FROM chat_message " +
        "WHERE conversation_id = ? " +
        "ORDER BY seq DESC " +
        "LIMIT ?";
    private static final String READ_CHAT_MESSAGES_BEFORE_SEQ_SQL =
        "SELECT seq, payload_json FROM chat_message " +
        "WHERE conversation_id = ? AND seq < ? " +
        "ORDER BY seq DESC " +
        "LIMIT ?";

    private static final Set<String> ALLOWED_EXECUTE_SQL = new HashSet<>(Arrays.asList(
        normalizeSql(CREATE_ENTRIES_TABLE_SQL),
        normalizeSql(CREATE_CHAT_CONVERSATION_TABLE_SQL),
        normalizeSql(CREATE_CHAT_MESSAGE_TABLE_SQL),
        normalizeSql(CREATE_CHAT_CONVERSATION_UPDATED_INDEX_SQL),
        normalizeSql(CREATE_CHAT_MESSAGE_CONVERSATION_SEQ_INDEX_SQL),
        "BEGIN IMMEDIATE",
        "COMMIT",
        "ROLLBACK",
        normalizeSql(UPSERT_ENTRY_SQL),
        normalizeSql(DELETE_ENTRY_SQL),
        normalizeSql(UPSERT_CHAT_CONVERSATION_SQL),
        normalizeSql(DELETE_CHAT_CONVERSATION_MESSAGES_SQL),
        normalizeSql(UPSERT_CHAT_MESSAGE_SQL)
    ));
    private static final Set<String> ALLOWED_QUERY_SQL = new HashSet<>(Arrays.asList(
        normalizeSql(READ_ENTRY_SQL),
        normalizeSql(LIST_KEYS_WITH_PREFIX_SQL),
        normalizeSql(READ_CHAT_CONVERSATION_SUMMARIES_SQL),
        normalizeSql(READ_CHAT_CONVERSATION_EXISTS_SQL),
        normalizeSql(READ_CHAT_CONVERSATION_METADATA_SQL),
        normalizeSql(READ_CHAT_MESSAGE_COUNT_SQL),
        normalizeSql(READ_RECENT_CHAT_MESSAGES_SQL),
        normalizeSql(READ_CHAT_MESSAGES_BEFORE_SEQ_SQL)
    ));

    @PluginMethod
    public void execute(PluginCall call) {
        String sql = call.getString("sql");
        if (sql == null) {
            call.reject("缺少 LocalData SQLite SQL。");
            return;
        }
        JSArray params = call.getArray("params", new JSArray());

        sqliteExecutor.execute(() -> {
            try {
                String normalizedSql = normalizeSql(sql);
                if (!ALLOWED_EXECUTE_SQL.contains(normalizedSql)) {
                    throw new LocalDataSqliteException("LocalData SQLite 只允许仓库适配层声明的 SQL。");
                }
                SQLiteDatabase db = openDatabase();
                runStatement(db, normalizedSql, params);
                call.resolve();
            } catch (Exception error) {
                call.reject(failureMessage("执行 LocalData SQLite 写入失败", error), error);
            }
        });
    }

    @PluginMethod
    public void query(PluginCall call) {
        String sql = call.getString("sql");
        if (sql == null) {
            call.reject("缺少 LocalData SQLite SQL。");
            return;
        }
        JSArray params = call.getArray("params", new JSArray());

        sqliteExecutor.execute(() -> {
            try {
                String normalizedSql = normalizeSql(sql);
                if (!ALLOWED_QUERY_SQL.contains(normalizedSql)) {
                    throw new LocalDataSqliteException("LocalData SQLite 只允许仓库适配层声明的 SQL。");
                }
                SQLiteDatabase db = openDatabase();
                JSObject result = new JSObject();
                result.put("rows", runQuery(db, normalizedSql, params));
                call.resolve(result);
            } catch (Exception error) {
                call.reject(failureMessage("执行 LocalData SQLite 读取失败", error), error);
            }
        });
    }

    @Override
    protected void handleOnDestroy() {
        super.handleOnDestroy();
        if (database != null) {
            database.close();
            database = null;
        }
        sqliteExecutor.shutdown();
    }

    private SQLiteDatabase openDatabase() {
        if (database != null && database.isOpen()) {
            return database;
        }
        File directory = new File(getContext().getFilesDir(), "PolarisLocalDataSqlite");
        if (!directory.exists() && !directory.mkdirs()) {
            throw new LocalDataSqliteException("无法创建 LocalData SQLite 目录。");
        }
        File file = new File(directory, "local-data.sqlite3");
        database = SQLiteDatabase.openOrCreateDatabase(file, null);
        database.setForeignKeyConstraintsEnabled(true);
        return database;
    }

    private void runStatement(SQLiteDatabase db, String sql, JSArray params) throws Exception {
        if ("BEGIN IMMEDIATE".equals(sql) || "COMMIT".equals(sql) || "ROLLBACK".equals(sql)) {
            db.execSQL(sql);
            return;
        }
        SQLiteStatement statement = db.compileStatement(sql);
        try {
            bindParams(statement, params);
            statement.execute();
        } finally {
            statement.close();
        }
    }

    private JSONArray runQuery(SQLiteDatabase db, String sql, JSArray params) throws Exception {
        JSONArray rows = new JSONArray();
        String[] selectionArgs = queryArgs(params);
        try (Cursor cursor = db.rawQuery(sql, selectionArgs)) {
            while (cursor.moveToNext()) {
                JSONObject row = new JSONObject();
                for (int index = 0; index < cursor.getColumnCount(); index += 1) {
                    row.put(cursor.getColumnName(index), cursorValue(cursor, index));
                }
                rows.put(row);
            }
        }
        return rows;
    }

    private void bindParams(SQLiteStatement statement, JSArray params) throws Exception {
        for (int index = 0; index < params.length(); index += 1) {
            Object value = params.get(index);
            int bindIndex = index + 1;
            if (value == null || value == JSONObject.NULL) {
                statement.bindNull(bindIndex);
            } else if (value instanceof Boolean) {
                statement.bindLong(bindIndex, Boolean.TRUE.equals(value) ? 1 : 0);
            } else if (value instanceof Integer || value instanceof Long) {
                statement.bindLong(bindIndex, ((Number) value).longValue());
            } else if (value instanceof Number) {
                statement.bindDouble(bindIndex, ((Number) value).doubleValue());
            } else if (value instanceof String) {
                statement.bindString(bindIndex, (String) value);
            } else {
                throw new LocalDataSqliteException("LocalData SQLite 参数格式不正确。");
            }
        }
    }

    private String[] queryArgs(JSArray params) throws Exception {
        String[] args = new String[params.length()];
        for (int index = 0; index < params.length(); index += 1) {
            Object value = params.get(index);
            if (value == null || value == JSONObject.NULL) {
                args[index] = null;
            } else if (value instanceof String) {
                args[index] = (String) value;
            } else if (value instanceof Number || value instanceof Boolean) {
                args[index] = String.valueOf(value);
            } else {
                throw new LocalDataSqliteException("LocalData SQLite 参数格式不正确。");
            }
        }
        return args;
    }

    private Object cursorValue(Cursor cursor, int index) {
        switch (cursor.getType(index)) {
            case Cursor.FIELD_TYPE_NULL:
                return JSONObject.NULL;
            case Cursor.FIELD_TYPE_INTEGER:
                return cursor.getLong(index);
            case Cursor.FIELD_TYPE_FLOAT:
                return cursor.getDouble(index);
            case Cursor.FIELD_TYPE_BLOB:
                return Base64.encodeToString(cursor.getBlob(index), Base64.NO_WRAP);
            case Cursor.FIELD_TYPE_STRING:
            default:
                return cursor.getString(index);
        }
    }

    private static String normalizeSql(String sql) {
        return sql
            .trim()
            .replaceAll("\\s+", " ")
            .replaceAll("\\s*([(),=])\\s*", "$1");
    }

    private static String failureMessage(String prefix, Exception error) {
        String detail = error.getMessage();
        if (detail == null || detail.trim().isEmpty()) {
            return String.format(Locale.ROOT, "%s。", prefix);
        }
        return String.format(Locale.ROOT, "%s：%s", prefix, detail.trim());
    }

    private static class LocalDataSqliteException extends RuntimeException {
        LocalDataSqliteException(String message) {
            super(message);
        }
    }
}

File: ios/App/CapApp-SPM/Sources/CapApp-SPM/LocalDataSqlitePlugin.swift
import Capacitor
import Foundation
import SQLite3

private let localDataSqliteTransient = unsafeBitCast(-1, to: sqlite3_destructor_type.self)

@objc(LocalDataSqlitePlugin)
public class LocalDataSqlitePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "LocalDataSqlitePlugin"
    public let jsName = "LocalDataSqlite"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "execute", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "query", returnType: CAPPluginReturnPromise)
    ]

    private let sqliteQueue = DispatchQueue(label: "com.polaris.localData.sqlite", qos: .userInitiated)
    private var database: OpaquePointer?

    private let createEntriesTableSql = """
    CREATE TABLE IF NOT EXISTS local_data_entries (
      key TEXT PRIMARY KEY NOT NULL,
      value_json TEXT NOT NULL,
      updated_at INTEGER NOT NULL
    )
    """
    private let readEntrySql = """
    SELECT value_json
    FROM local_data_entries
    WHERE key = ?
    LIMIT 1
    """
    private let listKeysWithPrefixSql = """
    SELECT key
    FROM local_data_entries
    WHERE substr(key, 1, ?) = ?
    """
    private let upsertEntrySql = """
    INSERT INTO local_data_entries (key, value_json, updated_at)
    VALUES (?, ?, ?)
    ON CONFLICT(key) DO UPDATE SET
      value_json = excluded.value_json,
      updated_at = excluded.updated_at
    """
    private let deleteEntrySql = """
    DELETE FROM local_data_entries
    WHERE key = ?
    """
    private let createChatConversationTableSql = """
    CREATE TABLE IF NOT EXISTS chat_conversation (
      id TEXT PRIMARY KEY NOT NULL,
      title TEXT NOT NULL,
      kind TEXT NOT NULL,
      collaborator_id TEXT,
      group_room_id TEXT,
      active_project_id TEXT,
      pinned_at INTEGER,
      created_at INTEGER NOT NULL,
      updated_at INTEGER NOT NULL,
      metadata_json TEXT NOT NULL
    )
    """
    private let createChatMessageTableSql = """
    CREATE TABLE IF NOT EXISTS chat_message (
      id TEXT PRIMARY KEY NOT NULL,
      conversation_id TEXT NOT NULL,
      seq INTEGER NOT NULL,
      role TEXT NOT NULL,
      content TEXT NOT NULL,
      reasoning TEXT NOT NULL,
      created_at INTEGER NOT NULL,
      updated_at INTEGER NOT NULL,
      payload_json TEXT NOT NULL,
      FOREIGN KEY(conversation_id) REFERENCES chat_conversation(id) ON DELETE CASCADE,
      UNIQUE(conversation_id, seq)
    )
    """
    private let createChatConversationUpdatedIndexSql = """
    CREATE INDEX IF NOT EXISTS idx_chat_conversation_updated_at
    ON chat_conversation(updated_at DESC, id ASC)
    """
    private let createChatMessageConversationSeqIndexSql = """
    CREATE INDEX IF NOT EXISTS idx_chat_message_conversation_seq
    ON chat_message(conversation_id, seq)
    """
    private let upsertChatConversationSql = """
    INSERT INTO chat_conversation (
      id,
      title,
      kind,
      collaborator_id,
      group_room_id,
      active_project_id,
      pinned_at,
      created_at,
      updated_at,
      metadata_json
    ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    ON CONFLICT(id) DO UPDATE SET
      title = excluded.title,
      kind = excluded.kind,
      collaborator_id = excluded.collaborator_id,
      group_room_id = excluded.group_room_id,
      active_project_id = excluded.active_project_id,
      pinned_at = excluded.pinned_at,
      created_at = excluded.created_at,
      updated_at = excluded.updated_at,
      metadata_json = excluded.metadata_json
    """
    private let deleteChatConversationMessagesSql = """
    DELETE FROM chat_message
    WHERE conversation_id = ?
    """
    private let upsertChatMessageSql = """
    INSERT INTO chat_message (
      id,
      conversation_id,
      seq,
      role,
      content,
      reasoning,
      created_at,
      updated_at,
      payload_json
    ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
    ON CONFLICT(id) DO UPDATE SET
      conversation_id = excluded.conversation_id,
      seq = excluded.seq,
      role = excluded.role,
      content = excluded.content,
      reasoning = excluded.reasoning,
      created_at = excluded.created_at,
      updated_at = excluded.updated_at,
      payload_json = excluded.payload_json
    """
    private let readChatConversationSummariesSql = """
    SELECT
      c.id,
      c.title,
      c.kind,
      c.collaborator_id,
      c.group_room_id,
      c.active_project_id,
      c.pinned_at,
      c.created_at,
      c.updated_at,
      COUNT(m.id) AS message_count,
      COALESCE(MAX(m.updated_at), 0) AS latest_message_timestamp
    FROM chat_conversation c
    LEFT JOIN chat_message m ON m.conversation_id = c.id
    GROUP BY
      c.id,
      c.title,
      c.kind,
      c.collaborator_id,
      c.group_room_id,
      c.active_project_id,
      c.pinned_at,
      c.created_at,
      c.updated_at
    ORDER BY c.updated_at DESC, c.id ASC
    """
    private let readChatConversationExistsSql = """
    SELECT id
    FROM chat_conversation
    WHERE id = ?
    LIMIT 1
    """
    private let readChatConversationMetadataSql = """
    SELECT metadata_json
    FROM chat_conversation
    WHERE id = ?
    LIMIT 1
    """
    private let readChatMessageCountSql = """
    SELECT COUNT(*) AS message_count
    FROM chat_message
    WHERE conversation_id = ?
    """
    private let readRecentChatMessagesSql = """
    SELECT seq, payload_json
    FROM chat_message
    WHERE conversation_id = ?
    ORDER BY seq DESC
    LIMIT ?
    """
    private let readChatMessagesBeforeSeqSql = """
    SELECT seq, payload_json
    FROM chat_message
    WHERE conversation_id = ? AND seq < ?
    ORDER BY seq DESC
    LIMIT ?
    """

    deinit {
        if let database {
            sqlite3_close(database)
        }
    }

    @objc public func execute(_ call: CAPPluginCall) {
        guard let sql = call.getString("sql") else {
            call.reject("缺少 LocalData SQLite SQL。")
            return
        }
        let params = readParams(call)

        sqliteQueue.async {
            do {
                let normalizedSql = self.normalizeSql(sql)
                guard self.allowedExecuteSql().contains(normalizedSql) else {
                    throw LocalDataSqliteError.disallowedSql
                }
                let database = try self.openDatabase()
                try self.runStatement(database: database, sql: sql, params: params)
                DispatchQueue.main.async {
                    call.resolve()
                }
            } catch {
                DispatchQueue.main.async {
                    call.reject(self.failureMessage("执行 LocalData SQLite 写入失败", error), nil, error)
                }
            }
        }
    }

    @objc public func query(_ call: CAPPluginCall) {
        guard let sql = call.getString("sql") else {
            call.reject("缺少 LocalData SQLite SQL。")
            return
        }
        let params = readParams(call)

        sqliteQueue.async {
            do {
                guard self.allowedQuerySql().contains(self.normalizeSql(sql)) else {
                    throw LocalDataSqliteError.disallowedSql
                }
                let database = try self.openDatabase()
                let rows = try self.runQuery(database: database, sql: sql, params: params)
                DispatchQueue.main.async {
                    call.resolve(["rows": rows])
                }
            } catch {
                DispatchQueue.main.async {
                    call.reject(self.failureMessage("执行 LocalData SQLite 读取失败", error), nil, error)
                }
            }
        }
    }

    private func allowedExecuteSql() -> Set<String> {
        [
            normalizeSql(createEntriesTableSql),
            normalizeSql(createChatConversationTableSql),
            normalizeSql(createChatMessageTableSql),
            normalizeSql(createChatConversationUpdatedIndexSql),
            normalizeSql(createChatMessageConversationSeqIndexSql),
            "BEGIN IMMEDIATE",
            "COMMIT",
            "ROLLBACK",
            normalizeSql(upsertEntrySql),
            normalizeSql(deleteEntrySql),
            normalizeSql(upsertChatConversationSql),
            normalizeSql(deleteChatConversationMessagesSql),
            normalizeSql(upsertChatMessageSql)
        ]
    }

    private func allowedQuerySql() -> Set<String> {
        [
            normalizeSql(readEntrySql),
            normalizeSql(listKeysWithPrefixSql),
            normalizeSql(readChatConversationSummariesSql),
            normalizeSql(readChatConversationExistsSql),
            normalizeSql(readChatConversationMetadataSql),
            normalizeSql(readChatMessageCountSql),
            normalizeSql(readRecentChatMessagesSql),
            normalizeSql(readChatMessagesBeforeSeqSql)
        ]
    }

    private func readParams(_ call: CAPPluginCall) -> [Any] {
        guard let params = call.options["params"] as? [Any] else {
            return []
        }
        return params
    }

    private func openDatabase() throws -> OpaquePointer {
        if let database {
            return database
        }

        let directory = try databaseDirectory()
        let url = directory.appendingPathComponent("local-data.sqlite3")
        var openedDatabase: OpaquePointer?
        let flags = SQLITE_OPEN_CREATE | SQLITE_OPEN_READWRITE | SQLITE_OPEN_FULLMUTEX
        guard sqlite3_open_v2(url.path, &openedDatabase, flags, nil) == SQLITE_OK,
              let openedDatabase else {
            let message = openedDatabase.map { sqliteErrorMessage($0) } ?? "unknown sqlite open failure"
            if let openedDatabase {
                sqlite3_close(openedDatabase)
            }
            throw LocalDataSqliteError.sqlite(message)
        }

        database = openedDatabase
        return openedDatabase
    }

    private func databaseDirectory() throws -> URL {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
            ?? FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first
            ?? FileManager.default.temporaryDirectory
        let directory = base.appendingPathComponent("PolarisLocalDataSqlite", isDirectory: true)
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        return directory
    }

    private func runStatement(database: OpaquePointer, sql: String, params: [Any]) throws {
        var statement: OpaquePointer?
        try prepare(database: database, sql: sql, statement: &statement)
        defer {
            sqlite3_finalize(statement)
        }
        try bind(params: params, statement: statement)
        let result = sqlite3_step(statement)
        guard result == SQLITE_DONE else {
            throw LocalDataSqliteError.sqlite(sqliteErrorMessage(database))
        }
    }

    private func runQuery(database: OpaquePointer, sql: String, params: [Any]) throws -> [[String: Any]] {
        var statement: OpaquePointer?
        try prepare(database: database, sql: sql, statement: &statement)
        defer {
            sqlite3_finalize(statement)
        }
        try bind(params: params, statement: statement)

        var rows: [[String: Any]] = []
        while true {
            let result = sqlite3_step(statement)
            if result == SQLITE_DONE {
                return rows
            }
            guard result == SQLITE_ROW else {
                throw LocalDataSqliteError.sqlite(sqliteErrorMessage(database))
            }

            var row: [String: Any] = [:]
            for index in 0..<sqlite3_column_count(statement) {
                guard let namePointer = sqlite3_column_name(statement, index) else {
                    continue
                }
                let name = String(cString: namePointer)
                row[name] = sqliteColumnValue(statement: statement, index: index)
            }
            rows.append(row)
        }
    }

    private func prepare(database: OpaquePointer, sql: String, statement: inout OpaquePointer?) throws {
        guard sqlite3_prepare_v2(database, sql, -1, &statement, nil) == SQLITE_OK else {
            throw LocalDataSqliteError.sqlite(sqliteErrorMessage(database))
        }
    }

    private func bind(params: [Any], statement: OpaquePointer?) throws {
        for (offset, value) in params.enumerated() {
            let index = Int32(offset + 1)
            if value is NSNull {
                sqlite3_bind_null(statement, index)
                continue
            }
            if let stringValue = value as? String {
                sqlite3_bind_text(statement, index, stringValue, -1, localDataSqliteTransient)
                continue
            }
            if let boolValue = value as? Bool {
                sqlite3_bind_int(statement, index, boolValue ? 1 : 0)
                continue
            }
            if let intValue = value as? Int {
                sqlite3_bind_int64(statement, index, sqlite3_int64(intValue))
                continue
            }
            if let int64Value = value as? Int64 {
                sqlite3_bind_int64(statement, index, sqlite3_int64(int64Value))
                continue
            }
            if let doubleValue = value as? Double {
                sqlite3_bind_double(statement, index, doubleValue)
                continue
            }
            if let numberValue = value as? NSNumber {
                sqlite3_bind_double(statement, index, numberValue.doubleValue)
                continue
            }
            throw LocalDataSqliteError.invalidParams
        }
    }

    private func sqliteColumnValue(statement: OpaquePointer?, index: Int32) -> Any {
        switch sqlite3_column_type(statement, index) {
        case SQLITE_INTEGER:
            return NSNumber(value: sqlite3_column_int64(statement, index))
        case SQLITE_FLOAT:
            return NSNumber(value: sqlite3_column_double(statement, index))
        case SQLITE_TEXT:
            guard let textPointer = sqlite3_column_text(statement, index) else {
                return ""
            }
            return String(cString: textPointer)
        case SQLITE_NULL:
            return NSNull()
        case SQLITE_BLOB:
            let byteCount = Int(sqlite3_column_bytes(statement, index))
            guard let bytes = sqlite3_column_blob(statement, index), byteCount > 0 else {
                return ""
            }
            return Data(bytes: bytes, count: byteCount).base64EncodedString()
        default:
            return NSNull()
        }
    }

    private func normalizeSql(_ sql: String) -> String {
        sql
            .split(whereSeparator: { $0.isWhitespace })
            .joined(separator: " ")
            .replacingOccurrences(
                of: "\\s*([(),=])\\s*",
                with: "$1",
                options: .regularExpression
            )
    }

    private func sqliteErrorMessage(_ database: OpaquePointer) -> String {
        guard let message = sqlite3_errmsg(database) else {
            return "unknown sqlite failure"
        }
        return String(cString: message)
    }

    private func failureMessage(_ prefix: String, _ error: Error) -> String {
        let detail = (error as NSError).localizedDescription.trimmingCharacters(in: .whitespacesAndNewlines)
        if detail.isEmpty {
            return "\(prefix)。"
        }
        return "\(prefix)：\(detail)"
    }

    private enum LocalDataSqliteError: LocalizedError {
        case disallowedSql
        case invalidParams
        case sqlite(String)

        var errorDescription: String? {
            switch self {
            case .disallowedSql:
                return "LocalData SQLite 只允许仓库适配层声明的 SQL。"
            case .invalidParams:
                return "LocalData SQLite 参数格式不正确。"
            case .sqlite(let message):
                return message
            }
        }
    }
}
